### PR TITLE
[FLINK-8144] [table] Optimize the timer logic in RowTimeUnboundedOver

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request optimize the timer handling in RowTimeUnboundedOver. Currently the MapState will be scanned a lot of times if the watermark arrives some seconds later than the event.*

## Verifying this change

This change is already covered by existing tests, such as *OverWindowHarnessTest.testRowTimeUnboundedRangeOver*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowTimeUnboundedOver.scala [hunk 1]
```diff
@@ -116,7 +116,7 @@ abstract class RowTimeUnboundedOver(
     // discard late record
     if (timestamp > curWatermark) {
       // ensure every key just registers one timer
-      ctx.timerService.registerEventTimeTimer(curWatermark + 1)
+      ctx.timerService.registerEventTimeTimer(timestamp)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowTimeUnboundedOver.scala]: Registering multiple timers on the same timestamp (`curWatermark + 1`), means that only one timer exists that fires exactly once when a watermark is received.

By registering timers on different timestamps, we have many timers that all will fire when a watermark is received. I don't think this is what we want.

**Comment by dianfu** [comment]
{{AbstractKeyedCEPPatternOperator}} has similar logic as {{RowTimeUnboundedOver}}. As described in FLINK-8106, we find that the performance is very bad under the current logic for {{AbstractKeyedCEPPatternOperator}}. The throughput can increase from 10+ tps to about 3500 tps for one operator in the case of RocksDBStateBackend after optimizing the timer logic. I think the optimization should also apply to {{RowTimeUnboundedOver}}.

BTW: the watermark we use in the CEP use case is {{AssignerWithPunctuatedWatermarks}}. It will generate one watermark for every input element.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowTimeUnboundedOver.scala (hunk 1 above)
Comment (dianfu) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowTimeUnboundedOver.scala]: @fhueske Thanks a lot for your comments. Your concern makes sense to me. I think the current implementation is ok under periodic watermark. But I'm not sure if it's optimal under punctuated watermark. We will perform some performance test for unbounded over under punctuated watermark and share the results.

**Comment by fhueske** [comment]
I see. The bad performance is related to way that you generate the watermarks. Generating so many watermarks is very inefficient because you processing many more watermarks than actual records (watermarks are always broadcasted).
I'm not convinced that this is a scenario that we should optimize for.

**Comment by dianfu** [comment]
@fhueske  Yes, the bad performance is related to the way we generate the watermarks. But this is the way how punctuated watermark works. If we do not optimize for punctuated watermark, then are we suggesting users not to use punctuated watermark for unbounded over window?

For the side effect `By registering timers on different timestamps, we have many timers that all will fire when a watermark is received.`, I think we can add a variable such as `lastProcessedWatermark` to record the last processed watermark. Then there will be no much overhead. Thoughts?

**Comment by fhueske** [comment]
A punctuated watermark assigner can emit `null` instead of a watermark. So using a punctuated watermark assigner does not mean that for each event a watermark must be emitted. The motivation for punctuated watermarks is to react to specific events that carry watermark information and not to all events (see [JavaDocs](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/AssignerWithPunctuatedWatermarks.java)). Of course there's nothing that prevents users from emitting one watermark per event but as I said this was not the intended use case.

A `lastProcessedWatermark` state variable would improve the situation a bit by acting as an early out for the `onTimer()` method. However, the system would still need to checkpoint and handle significantly more timers than with the current implementation.

**Comment by dianfu** [comment]
@fhueske Thanks a lot for your comments. Definitely agree with you on the motivation of punctuated watermark. But in practice, there are many cases there is no watermark information in the source data at all. So we have to generate one watermark for every incoming event.

**Comment by fhueske** [comment]
I assume you had also tried periodic watermarks in your setting. I'm curious why they didn't work for you. 

A periodic watermark assigner extracts the timestamp from each record and just emits a watermark when it is asked. From a correctness point of view, this should be the same as punctuated watermarks, just with lower overhead and higher watermark latency.

**Comment by dianfu** [comment]
For historical reasons, we only use punctuated watermark currently. But I think you are right, we should consider using periodic watermark instead. Thanks a lot for your explanation. I will close this PR.

**Comment by fhueske** [comment]
Thank you @dianfu for the discussion. 
I raised concerns about this PR because it seem to me to optimize for a special use case and possibly cause regression for more common workloads. However, I don't have numbers to back my assumptions.

If this is a big issue for you, we can also benchmark the impact of the change for periodic and punctuated watermarks to check how it affects the performance.

**Comment by dianfu** [comment]
The current known impact is that it will need to checkpoint more timer. For the impact to performance, I will do some benchmark later. :)   
I think we only need to consider this optimization if there are cases where users have to use punctuated watermark in the way that generating one watermark for every incoming event. I will reopen this PR if we find a persuading use case and reason. :)

**Comment by fhueske** [comment]
Great, thank you!

**Pull request closed without merging**